src/test/rgw/test_d4n_filter.cc

```diff
@@ -0,0 +1,2810 @@
+
+    {
+      boost::system::error_code ec;
+      request req;
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName);
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", TEST_BUCKET + testName + "__:null_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", TEST_BUCKET + testName + "_" + TEST_OBJ + testName + "_0_9");
+      req.push("FLUSHALL");
+
+      response< int, int, int, int, boost::redis::ignore_t > resp;
+
+      conn->async_exec(req, resp, yield[ec]);
+      ASSERT_EQ((bool)ec, false);
+      EXPECT_EQ((int)std::get<0>(resp).value(), 0);
+      EXPECT_EQ((int)std::get<1>(resp).value(), 0);
+      EXPECT_EQ((int)std::get<2>(resp).value(), 0);
+      EXPECT_EQ((int)std::get<3>(resp).value(), 0);
+    }
+
+    std::error_code err;
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + version, err), false);  
+    std::string oid = version + "#0#" + std::to_string(ofs);
+    EXPECT_EQ(fs::exists(CACHE_DIR + "/" + TEST_BUCKET + testName + "/" + TEST_OBJ + testName + "/" + oid, err), false);     
+
+    conn->cancel();
+    testBucket->remove(env->dpp, true, optional_yield{yield});
+    DriverDestructor driver_destructor(static_cast<rgw::sal::D4NFilterDriver*>(driver));
+  }, rethrow);
+
+  io.run();
+}
+
+// Write cache tests, versioned
+TEST_F(D4NFilterFixture, PutVersionedObjectWrite)
+{
+  env->cct->_conf->d4n_writecache_enabled = true;
+  env->cct->_conf->rgw_d4n_cache_cleaning_interval = 2;
+  const std::string testName = "PutVersionedObjectWrite";
+  const std::string bucketName = "/tmp/d4n_filter_tests/dbstore-default_ns.1";
+  std::string version, instance;
+ 
+  boost::asio::spawn(io, [this, &testName, &bucketName, &version, &instance] (boost::asio::yield_context yield) {
+    init_driver(yield);
+    create_bucket(testName, yield);
+    // Overwrite testName because the cleaning method derives the bucket name differently from other ops.
+    testBucket->get_info().bucket.bucket_id = "/tmp/d4n_filter_tests/dbstore-default_ns.1";
+    put_version_enabled_object(testName, instance, yield);
+    std::this_thread::sleep_for(std::chrono::seconds(1)); // So ZADD calls set values in the same order
+    put_version_suspended_object(testName, yield);
+
+    {
+      boost::system::error_code ec;
+      request req;
+      req.push("EXISTS", bucketName + "_" + TEST_OBJ + testName);
+      req.push("EXISTS", bucketName + "_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", bucketName + "_" + TEST_OBJ + testName + "_0_9");
+      req.push("EXISTS", bucketName + "__:" + instance + "_" + TEST_OBJ + testName + "_0_0");
+      req.push("EXISTS", bucketName + "__:" + instance + "_" + TEST_OBJ + testName + "_0_9");
+      req.push("EXISTS", bucketName + "__:null_"  + TEST_OBJ + testName + "_0_0");
+      req.push("ZRANGE", bucketName + "_" + TEST_OBJ + testName, "0", "10000000000");
```

Comment:
ordered set stores elements in ascending order, and we store versions based on the creation time, so the latest version will be saved last in the set. You can use zrevrange to get the versions and the top element must be the latest version which is "null" and then the next version will be the instance version. If you use zrange, the order will be reversed.